### PR TITLE
Added promptpay component

### PR DESCRIPTION
**Pull request by neelSharma12**
This PR contains implementation of new component: PromptPay.

Pending:
1. Localization needs to be added.

### CODE REVIEW

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            onTick: { [weak self] in self?.updateExpiration($0) },
+            onExpiration: { [weak self] in self?.onTimerTimeout() }
+        )
+        timeoutTimer?.startTimer()
+    }
+    
+    private func updateExpiration(_ timeLeft: TimeInterval) {
+        progress.completedUnitCount = Int64(timeLeft)
+        let timeLeftString = timeLeft.adyen.timeLeftString() ?? ""
+        expirationText = localizedString(.pixExpirationLabel,
```
Comment (neelSharma12) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: need to change localizedString

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        let viewController = QRCodeViewController(viewModel: createModel(with: action))
+        viewController.qrCodeView.delegate = self
+        return viewController
+    }
+    
+    private func createModel(with action: QRCodeAction) -> QRCodeView.Model {
+        let url = LogoURLProvider.logoURL(withName: action.paymentMethodType.rawValue, environment: context.apiContext.environment)
+        return QRCodeView.Model(
+            action: action,
+            instruction: localizedString(.pixInstructions,
```
Comment (neelSharma12) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: need to change localizedString

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        self.expirationTimeout = timeoutInterval
+        
+        updateExpiration(timeoutInterval)
+    }
+
+    /// Handles QR code action.
+    ///
+    /// - Parameter action: The QR code action.
+    public func handle(_ action: QRCodeAction) {
+        assert(presentationDelegate != nil)
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: this is not needed because you already use `AdyenAssertion`, right?

File: AdyenActions/UI/View Controllers/QR Code/QRCodeViewController.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        
+        // swiftlint:disable:next unused_setter_value
+        set { AdyenAssertion.assertionFailure(message: """
+        PreferredContentSize is overridden for this view controller.
+        getter - returns minimum possible content size.
+        setter - no implemented.
+        """) }
+    }
+    
+}
```
Comment (mohammedDehairy) [inline AdyenActions/UI/View Controllers/QR Code/QRCodeViewController.swift]: this class is doing exactly like `PixViewController` except that its using `QRCodeView ` instead of `PixView`, can we generalize this `ViewController ` using generics or something?, or even reuse the `PixView ` and make it work for both `pix` and `QRCode`? they're very similar anyway, right?

File: AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        expirationLabel.accessibilityIdentifier = ViewIdentifierBuilder.build(scopeInstance: self, postfix: "expirationLabel")
+        
+        bind(model.expiration, to: expirationLabel, at: \.text)
+        
+        return expirationLabel
+    }()
+
+    @objc private func saveQRCodeAsImage() {
+        delegate?.saveAsImage(qrCodeImage: qrCodeImageView.adyen.snapShot(), sourceView: saveAsImageButton)
+    }
```
Comment (mohammedDehairy) [inline AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift:233]: most of this code is very similar to `PixView`, this seems like code duplication to me :-)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+@_spi(AdyenInternal)
+extension QRCodeActionComponent: QRCodeViewDelegate {
+
+    internal func saveAsImage(qrCodeImage: UIImage?, sourceView: UIView) {
+        presentSharePopover(with: qrCodeImage as Any, sourceView: sourceView)
+    }
+    
+}
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift:241]: this file is very similar to `PixComponent`, I am afraid there are quite a bit of code duplication here

File: AdyenActions/UI/View Controllers/QR Code/QRCodeViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            self.action = action
+            self.instruction = instruction
+            self.payment = payment
+            self.logoUrl = logoUrl
+            self.observedProgress = observedProgress
+            self.expiration = expiration
+            self.style = style
+        }
+        
+    }
```
Comment (mohammedDehairy) [inline AdyenActions/UI/View Controllers/QR Code/QRCodeViewModel.swift:77]: also this code seems like duplicated from `PixView.ViewModel` :-)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 8]
```diff
@@ -84,35 +84,42 @@ public final class PixActionComponent: ActionComponent, Cancellable {
...
 
     /// Handles QR code action.
     ///
     /// - Parameter action: The QR code action.
     public func handle(_ action: QRCodeAction) {
-        let pollingComponent = pollingComponentBuilder.handler(for: action.paymentMethodType)
-        pollingComponent.delegate = self
-        
-        assert(presentationDelegate != nil)
+        let pollingComponent = pollingComponentBuilder?.handler(for: action.paymentMethodType)
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: `pollingComponent` is not retained, it will get deallocated immediately after the function returns, :-)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 9]
```diff
@@ -175,12 +185,12 @@ public final class PixActionComponent: ActionComponent, Cancellable {
     
     fileprivate func cleanup() {
         timeoutTimer?.stopTimer()
-        pollingComponent?.didCancel()
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: this is still needed for PIX, right? why is this removed?

File: AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift [hunk 10]
```diff
@@ -130,6 +179,30 @@ internal final class PixView: UIView, Localizable, AdyenObserver {
         return instructionLabel
     }()
     
+    private lazy var qrCodeImageView: UIImageView = {
+        let data = model.action.qrCodeData.data(using: String.Encoding.ascii)
+        if let filter = CIFilter(name: "CIQRCodeGenerator") {
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift]: It's better to return the `UIImageView()` in a guard and return the actual object outside of the guard to make it more prominent

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 8 above)
Comment (erenbesel) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: This is a nice catch. But apparently it's been like this for a while. Has it not been working all this time? 🤔

File: AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift (hunk 10 above)
Comment (neelSharma12) [reply AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift]: Sure :)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 8 above)
Comment (mohammedDehairy) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: oh wow, shit, that's definitely a bug :-D

Comment (neelSharma12) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: fixed this bug. Thank you guys for identifying this. Teamwork makes the dream work 🕺

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 9 above)
Comment (neelSharma12) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: Added this back 😄

**Review by mohammedDehairy — DISMISSED** [review]

**Review by nauaros — DISMISSED** [review]

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 11]
```diff
@@ -132,9 +144,16 @@ public final class QRCodeActionComponent: ActionComponent, Cancellable {
     private func updateExpiration(_ timeLeft: TimeInterval) {
         progress.completedUnitCount = Int64(timeLeft)
         let timeLeftString = timeLeft.adyen.timeLeftString() ?? ""
-        expirationText = localizedString(.pixExpirationLabel,
-                                         configuration.localizationParameters,
-                                         timeLeftString)
+        guard let action = qrCodeAction else {return}
+        if action.paymentMethodType == .promptPay {
```
Comment (erenbesel) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: Could we use a switch for this type? I think there are more qrCode types on the way so let's be error-proof in advance :)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 12]
```diff
@@ -144,31 +163,43 @@ public final class QRCodeActionComponent: ActionComponent, Cancellable {
...
-        viewController.pixView.delegate = self
+        let viewController = QRCodeViewController(viewModel: createModel(with: action))
+        viewController.qrCodeView.delegate = self
         return viewController
     }
-    
-    private func createModel(with action: QRCodeAction) -> PixView.Model {
+
+    private func getQRCodeInstruction(with action: QRCodeAction) -> String {
+        if action.paymentMethodType == .promptPay {
```
Comment (erenbesel) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: same here (switch)

File: AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    }
+    
+    private func buildUI() {
+        addLogo()
+        addInstructionLabel()
+        addQRCodeImage()
+        addAmountToPay()
+        addProgressView()
+        addExpirationLabel()
+        if model.action.paymentMethodType == .pix {
```
Comment (erenbesel) [inline AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift]: Hmm so some types will have copy code and some will have save as image. This decision better be in the viewmodel instead of the view. 
Let's add a property on the model that decides whether to show copyCode or saveImage (maybe enum?)

**Review by erenbesel — DISMISSED** [review]

File: AdyenActions/Components/Await/PollingHandlerProvider.swift [hunk 14]
```diff
@@ -22,7 +22,7 @@ internal protocol AnyPollingHandlerProvider {
 
     func handler(for paymentMethodType: AwaitPaymentMethod) -> AnyPollingHandler
     
-    func handler(for qrPaymentMethodType: QRCodePaymentMethod) -> AnyPollingHandler
+    func handler(for qrPaymentMethodType: QRCodePaymentMethod) -> AnyPollingHandler?
```
Comment (mohammedDehairy) [inline AdyenActions/Components/Await/PollingHandlerProvider.swift]: why is this optional? :-)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 15]
```diff
@@ -28,7 +28,9 @@ public final class QRCodeActionComponent: ActionComponent, Cancellable {
     public weak var presentationDelegate: PresentationDelegate?
 
     public weak var delegate: ActionComponentDelegate?
-    
+
+    internal let presenterViewController = UIViewController()
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift:32]: what's the purpose of this viewController?

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 16]
```diff
@@ -51,15 +53,17 @@ public final class QRCodeActionComponent: ActionComponent, Cancellable {
...
+    private let pollingComponentBuilder: AnyPollingHandlerProvider?
     
     private var pollingComponent: AnyPollingHandler?
-    
+
     private var timeoutTimer: ExpirationTimer?
     
     private let progress = Progress()
     
+    private var qrCodeAction: QRCodeAction?
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: instead of keeping it as an instance property, maybe pass it as parameter to the functions that needed?

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift [hunk 17]
```diff
@@ -84,34 +88,42 @@ public final class QRCodeActionComponent: ActionComponent, Cancellable {
...
             )
-        )
+            presentationDelegate.present(component: presentableComponent)
+        } else {
+            AdyenAssertion.assertionFailure(
+                message: "PresentationDelegate is nil. Provide a presentation delegate to QRCodeActionComponent."
+            )
+        }
         
         startTimer()
```
Comment (mohammedDehairy) [inline AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: you can pass the action object as a parameter here instead of keeping it an instance property, what do you think? :-)

File: AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+    @objc private func saveQRCodeAsImage() {
+        delegate?.saveAsImage(qrCodeImage: qrCodeImageView.adyen.snapShot(), sourceView: saveAsImageButton)
+    }
+
+    @objc private func copyCode() {
+        delegate?.copyToPasteboard(with: model.action)
+    }
+
+}
```
Comment (mohammedDehairy) [inline AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift:239]: the code looks nice! :-)

File: UITests/Components/QRCode/QRCodeActionComponentUITests.swift [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+@_spi(AdyenInternal) @testable import Adyen
+@testable import AdyenActions
+import XCTest
+
+class QRCodeActionComponentUITests: XCTestCase {
```
Comment (mohammedDehairy) [inline UITests/Components/QRCode/QRCodeActionComponentUITests.swift:11]: Nice!

File: AdyenActions/Components/Await/PollingHandlerProvider.swift (hunk 14 above)
Comment (neelSharma12) [reply AdyenActions/Components/Await/PollingHandlerProvider.swift]: oh! I think it got messed up during conflict resolution. I will fix this.

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 15 above)
Comment (neelSharma12) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift:32]: As I am using ShareableComponent protocol, it asked me to add it to conform to this protocol.

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 16 above)
Comment (neelSharma12) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: sure.

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 17 above)
Comment (neelSharma12) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift]: yes.

File: AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift (hunk 18 above)
Comment (neelSharma12) [reply AdyenActions/UI/View Controllers/QR Code/QRCodeView.swift:239]: thanks :)

File: UITests/Components/QRCode/QRCodeActionComponentUITests.swift (hunk 19 above)
Comment (neelSharma12) [reply UITests/Components/QRCode/QRCodeActionComponentUITests.swift:11]: thanks :)

File: AdyenActions/Components/QRCode/QRCodeActionComponent.swift (hunk 15 above)
Comment (mohammedDehairy) [reply AdyenActions/Components/QRCode/QRCodeActionComponent.swift:32]: hmm, I see, ok 👍

**Review by erenbesel — APPROVED** [review]

**Review by mohammedDehairy — APPROVED** [review]
